### PR TITLE
Only execute rules once with multiphase evaluation

### PR DESCRIPTION
**Pull request by anuraaga**
@M4tteoP do you know of a simple test for this? Maybe can add a commit to this PR

### CODE REVIEW

File: internal/corazawaf/rule.go [hunk 1]
```diff
@@ -295,7 +307,7 @@ func (r *Rule) doEvaluate(phase types.RulePhase, tx *Transaction, cache map[tran
 
 		}
 		if r.ID_ != 0 {
-			// we avoid matching chains and secmarkers
+			// we avoid matching secmarkers
```
Comment (anuraaga) [inline internal/corazawaf/rule.go:310]: Not related to the PR but noticed while trying to grok the function. We already checked `ParentID_` above so I think this `ID_` check is only to handle the secmarker case, not chain case

**Comment by M4tteoP** [comment]
This PR follows https://github.com/corazawaf/coraza/pull/565#issuecomment-1381342702, but on a second thought, I'm not anymore sure about the expected behaviour.

Trying the following request against `modsec2-apache`: `curl 'http://localhost:80/⚠' -H "Ciao:⚠"`
I have the following logs:
```
[id "920274"] [msg "Invalid character in request headers (outside of very strict set)"] [data "REQUEST_HEADERS:Ciao=\\xe2\\x9a\\xa0"] [severity "CRITICAL"] [ver "OWASP_CRS/4.0.0-rc1"] [tag "modsecurity"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "OWASP_CRS"] [tag "capec/1000/210/272"] [tag "paranoia-level/4"] [hostname "localhost"] [uri "/%f0%9f%86%98"] [unique_id "Y-UG0dofrjeK_NUHj6QmYQAAAEM"]

[id "920272"] [msg "Invalid character in request (outside of printable chars below ascii 127)"] [data "REQUEST_URI=/\\xf0\\x9f\\x86\\x98"] [severity "CRITICAL"] [ver "OWASP_CRS/4.0.0-rc1"] [tag "modsecurity"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "OWASP_CRS"] [tag "capec/1000/210/272"] [tag "paranoia-level/3"] [hostname "localhost"] [uri "/\\xf0\\x9f\\x86\\x98"] [unique_id "Y-UG0dofrjeK_NUHj6QmYQAAAEM"]

[id "920272"] [msg "Invalid character in request (outside of printable chars below ascii 127)"] [data "REQUEST_HEADERS:Ciao=\\xe2\\x9a\\xa0"] [severity "CRITICAL"] [ver "OWASP_CRS/4.0.0-rc1"] [tag "modsecurity"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "OWASP_CRS"] [tag "capec/1000/210/272"] [tag "paranoia-level/3"] [hostname "localhost"] [uri "/\\xf0\\x9f\\x86\\x98"] [unique_id "Y-UG0dofrjeK_NUHj6QmYQAAAEM"]

[id "949110"] [msg "Inbound Anomaly Score Exceeded (Total Score: 15)"] [ver "OWASP_CRS/4.0.0-rc1"] [tag "modsecurity"] [tag "anomaly-evaluation"] [hostname "localhost"] [uri "/\\xf0\\x9f\\x86\\x98"] [unique_id "Y-UG0dofrjeK_NUHj6QmYQAAAEM"]

[id "980170"] [msg "Anomaly Scores: (Inbound Scores: blocking=15, detection=15, per_pl=0-0-10-5, threshold=5) - (Outbound Scores: blocking=0, detection=0, per_pl=0-0-0-0, threshold=4) - (SQLI=0, XSS=0, RFI=0, LFI=0, RCE=0, PHPI=0, HTTP=0, SESS=0, COMBINED_SCORE=15)"] [ver "OWASP_CRS/4.0.0-rc1"] [tag "modsecurity"] [tag "reporting"] [hostname "localhost"] [uri "/\\xf0\\x9f\\x86\\x98"] [unique_id "Y-UG0dofrjeK_NUHj6QmYQAAAEM"]
```
Rule `920272` (it has multiple targets (`REQUEST_URI|REQUEST_HEADERS|ARGS|ARGS_NAMES|REQUEST_BODY`)), has been triggered two times, and the `anomaly score` increased accordingly twice. The same happens in Coraza (Without taking into account the multiphase functionality).

So the anomaly score is incremented more than once for the same rule. 

If it is correct, speaking about multiphase we should be good like we are:
- each target is checked just once (at most it is anticipated to be a previous phase, but not checked twice)
- each rule can be triggered more than once against different targets

cc @theseion @airween could you help me to clarify this behaviour? 🙏

I wrote a [test](https://github.com/corazawaf/coraza/commit/eaf39e6335bca79026e59766dcc76e550b429718) and some tweaks attempt to make their PR work, but first I wish to be sure about the previous reasoning. Now, the test is working with multiphase, but not with standard Coraza

**Comment by theseion** [comment]
Ha, you're right! The rule is executed against each target. So as long as a rule is executed exactly once for each of its targets (thinking of multiphase), we should be good.

Unfortunately, the nginx connector seems to behave differently. It might just be the logger though, as the anomaly score looks to be the same.

```
# nginx
curl -H "x-crs-paranoia-level: 3" -H "x-format-output: txt-matched-rules" -H "x-backend: nginx" -H "x-crs-version: nightly" 'https://sandbox.coreruleset.org/⚠' -H "Ciao:⚠"
920272 PL3 Invalid character in request (outside of printable chars below ascii 127)
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 10)

# httpd
curl -H "x-crs-paranoia-level: 4" -H "x-format-output: txt-matched-rules" -H "x-backend: nginx" -H "x-crs-version: nightly" 'https://sandbox.coreruleset.org/⚠' -H "Ciao:⚠"
920272 PL3 Invalid character in request (outside of printable chars below ascii 127)
920272 PL3 Invalid character in request (outside of printable chars below ascii 127)
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 10)
980170 PL1 Anomaly Scores: (Inbound Scores: blocking=10, detection=10, per_pl=0-0-10-0, threshold=5) - (Outbound Scores: blocking=0, detection=0, per_pl=0-0-0-0, threshold=4) - (SQLI=0, XSS=0, RFI=0, LFI=0, RCE=0, PHPI=0, HTTP=0, SESS=0, COMBINED_SCORE=10)
```

**Comment by anuraaga** [comment]
Thanks for confirming this! I'll go ahead and close this PR then since it looks like we don't need it, @M4tteoP it will still be good to get your test in as a regression test, probably to CRS tests themselves rather than here since AFAICT we're not currently covering this case with FTW

**Pull request closed without merging**